OTAnalytics/plugin_parser/streaming_parser.py

```diff
@@ -0,0 +1,349 @@
+        self._tracks_dict: dict[TrackId, list[Detection]] = dict()
+
+    def parse_tracks(
+        self,
+        input_file: str,
+        detections: list[dict],
+        metadata_video: dict,
+        id_generator: TrackIdGenerator = TrackId,
+    ) -> Iterator[Track]:
+        for det_dict in detections:
+            det = parse_python_detection(
+                metadata_video, id_generator, det_dict, input_file
+            )
+
+            # Group detections by track id
+            if not self._tracks_dict.get(det.track_id):
+                self._tracks_dict[det.track_id] = []
+            self._tracks_dict[det.track_id].append(det)
+
+            # the finished flag indicates tha last detection of a track
+            # so the detections can be assembled to a track object
+            if det_dict[ottrk_format.FINISHED]:
+                track_detections = self._tracks_dict[det.track_id]
+                del self._tracks_dict[det.track_id]
+
+                track = create_python_track(
+                    det.track_id,
+                    track_detections,
+                    self._track_classification_calculator,
+                    self._track_length_limit,
+                )  # yield finished track
+                if track is not None:
+                    yield track
+
+    # after all files have been processed,
+    # yield all remaining tracks without finished flag
+    def get_remaining_tracks(self) -> Iterator[Track]:
+        for (
+            track_id,
+            detections,
+        ) in self._tracks_dict.items():
+            track = create_python_track(
+                track_id,
+                detections,
+                self._track_classification_calculator,
+                self._track_length_limit,
+            )  # yield remaining track
+            if track is not None:
+                yield track
+
+
+class StreamTrackParser(ABC):
+    @abstractmethod
+    def parse(self, files: set[Path]) -> Iterator[TrackDataset]:
+        """
+        Parse multiple track files and provide
+        the parsed Tracks in form of a lazy stream.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def register_tracks_metadata(self, tracks_metadata: TracksMetadata) -> None:
+        """Register TracksMetadata to be updated when a new ottrk file is parsed."""
+        raise NotImplementedError
+
+    @abstractmethod
+    def register_videos_metadata(self, videos_metadata: VideosMetadata) -> None:
+        """Register VideosMetadata to be updated when a new ottrk file is parsed."""
+        raise NotImplementedError
+
+
+TrackDatasetFactory = Callable[[list[Track]], TrackDataset]
+
+
+def default_track_dataset_factory(tracks: list[Track]) -> TrackDataset:
+    return PandasTrackDataset.from_list(
+        tracks=tracks,
+        track_geometry_factory=PygeosTrackGeometryDataset.from_track_dataset,
+        calculator=PandasByMaxConfidence(),
+    )
+
+
+class StreamOttrkParser(StreamTrackParser):
```

Comment:
This will be the future main parser for track data. Please add a unit test using our simple example in tests/data. The example has 60 frames. Please test the edge cases reading the first chunk of frames and the last chunk of frames.